### PR TITLE
food-chain: Rewrite tests and add working stub solution.

### PR DESCRIPTION
**Pull request by rbasso**
- Use lyrics from `x-common/exercises/food-chain/canonical-data.json`.
- Tests all lines, one by one, and then the entire song.
- Add stub solution passing the tests.
- Patch example solution to work with the new tests.

This is a first try at solving #301. When we get this right, we may extend the same logic to `beer-song` and `house`.

There is some magic in the test suite to check the lyrics line by line, but I couldn't find a better way to write it.

The stub solution has the complete lyrics in the most raw form possible, so that the student will be forced to refactor everything from zero.

The previous tests suite had an empty line at the end, but that makes no sense and doesn't match the reference file.

### CODE REVIEW

File: exercises/food-chain/HINTS.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+The challenge is to rewrite it until you are proud of it, and learn
+something in the process.
+
+If you don't know where to start, here are some ideas:
+
+- Try to reduce repetition to a minimum.
+- Try to make the code readable.
+- Generalize your solution, allowing the lady to swallow other things.
+
+Take you time.
```
Comment (petertseng) [inline exercises/food-chain/HINTS.md]: "Take your time"

Comment (rbasso) [inline exercises/food-chain/HINTS.md]: Fixed. :smile: 

Anything else?

**Comment by rbasso** [comment]
Thank you for the review, @petertseng! :smile: 

> Were you expecting more opposition?

A little, considering that this is the first exercise where we test two multi-line string, line by line.
Also, I have to admit that my custom *zipWith* function (`go`) is at least unusual. :smile: 

> ...keep in mind I haven't done this exercise in any language.

In that case I recommend you to try this one when you get the time. 

It may seem pointless at first, but I think I learn a few things refactoring it.

**Comment by petertseng** [comment]
I spent a little more time on ```go shouldBe `on` zip [1 :: Int ..] . lines``` than `go` since `go` was easy enough to understand. I'm not sure I could have done better than `go` since the goal is to stop at the longer of the lists rather than the shorter.

> In that case I recommend you to try this one when you get the time. 

I'll start now.

**Comment by petertseng** [comment]
Hmm, here's one thing that caught my eye. The failure message -

```
Failures:

  test/Tests.hs:27: 
  1) food-chain.song matches lines
       expected: Just (2,"I don't know why she swallowed the fly. Perhaps she'll die.")
        but got: Just (2,"Hmm, I messed up")
```

It may not be clear to the reader that this means "there was a mismatch at line 2", maybe?

**Comment by rbasso** [comment]
Yeah! I know... it is far from good.

The problem is that, to compare with a possibly missing line, we need to compare *Maybes*. I don't know how to customize an error message... I need to learn more about `hspec`.

About the line number, we can have it in the data compared, like we did, or in the test description. In the later case, it would create one test case for each line, and that creates a lot of output.

I agree that it deserves to be improved, but I still don't know exactly how. :confused:

**Pull request merged**